backtrack/permutation.go

```diff
@@ -20,22 +20,22 @@ func Permutation(nums []int) [][]int {
 	perm := make([]int, n)
 	vis := make([]bool, n)
 	var permutations [][]int
-	var backtrackPerm func(int)
-	backtrackPerm = func(depth int) {
-		if depth == n {
-			permutations = append(permutations, append([]int{}, perm...))
-			return
-		}
-		for i := range nums {
-			if vis[i] || (i > 0 && !vis[i-1] && nums[i] == nums[i-1]) {
-				continue
-			} // get first element that never used/visited
-			perm[depth] = nums[i]
-			vis[i] = true
-			backtrackPerm(depth + 1)
-			vis[i] = false
-		}
-	}
-	backtrackPerm(0)
+	backtrackPerm(&permutations, nums, 0, n, perm, vis)
 	return permutations
 }
+
+func backtrackPerm(permutations *[][]int, nums []int, depth int, n int, perm []int, vis []bool) {
```

Comment:
I don't think you need a pointer to append... I'll open a gitpod Saturday and see if in-fact we do indeed need it but `[][]int` is a slice (internally a pointer) so I'm under the impression that we don't need the pointer to it. But will let you know on Friday as soon as I get some time 😄

Reply:
Aren't slices passed by reference? Because they themselves are pointers to an underlying array, so I think we do not need to take a pointer.

Reply:
Its length cannot be changed, otherwise you will get a new slice.
Pre-calculation may not be a good solution as well, because it is probably a waste of space.
In some languages such as c, a top-level variable may be used here.
And the idea of some languages like java is to use classes here to organize the code, then it puts some variables that need to be passed as fields of the class.
Both of these can easily cause confusion for users.
Closures, on the other hand, achieve the effect of having the above inside a function, which is why I used it in the first place.But it has the disadvantage of not being clear enough.
The two function cases are somewhat complicated by the characteristics of go itself.
go's [][]int doesn't come with a resize method like c++'s vector. It is done through an external method.

Reply:
Ah, I see the problem. The function itself is modifying the slice. 👍🏽